lib/python/treadmill_aws/ipa525.py

```diff
@@ -0,0 +1,217 @@
+"""Implements IPA525 credential service.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import base64
+import grp
+import io
+import logging
+import json
+import os
+import pwd
+import re
+import shutil
+import tempfile
+
+from twisted.internet import reactor
+from twisted.internet import protocol
+
+from treadmill import gssapiprotocol
+from treadmill import utils
+
+_LOGGER = logging.getLogger(__name__)
+
+
+# Disable warning for too many branches.
+# pylint: disable=R0912
+def run_server(port, client_realm, server_realm, admin_group):
+    """Runs IPA525 server."""
+    _LOGGER.info(
+        'IPA 525 server starting - listening on port %d',
+        port)
+
+    def _parse_name(princ):
+        match = re.search('^([^/@]*)/?([^@]*)@?(.*)$', princ)
+        if match is None:
+            _LOGGER.error(
+                'Internal error - expected name to be a kerberos '
+                'principal name, got [%s]', princ)
+            raise ValueError('Internal error')
+
+        name = match.group(1)
+        inst = match.group(2)
+        realm = match.group(3)
+
+        return [name, inst, realm]
+
+    class IPA525Server(gssapiprotocol.GSSAPILineServer):
```

Comment:
consider creating a base class from GSSAPILineServer (or add methods to GSSAPILineServer) - there is a lot of repeatable code in three PRs.